### PR TITLE
Add data loader to GraphData class

### PR DESCRIPTION
**Pull request by MisterPNP**
The previous GraphData only contained is_match, a function allowing to differentiate between tabular and graph CSV datasets.

This PR adds the functionality of reading/loading the graph data from the input CSV into a NetworkX graph. This graph contains nodes, edges, and possible attributes for both. The use of NetworkX is very useful for downstream profiling of the data due to NetworkX's integrated functions to analyze and pull information from graphs.

### CODE REVIEW

**Review by JGSweets — CHANGES_REQUESTED** [review]
Changes requested from previous closed PR #527 . Will unblock once updated.

File: dataprofiler/data_readers/graph_data.py [hunk 1]
```diff
@@ -5,15 +5,73 @@
...
+        self._destination_node = options.get("destination_node", None)
+        self._target_keywords = options.get("target_keywords", ['target', 'destination', 'dst'])
+        self._source_keywords = options.get("source_keywords", ['source', 'src', 'origin'])
+        self._column_names = options.get("column_names", self.csv_column_names(self.input_file_path, self.options))
+
+        self._delimiter = options.get("delimiter", None)
+        self._quotechar = options.get("quotechar", None)
+        self._header = options.get("header", 'auto')
+
+        self._data = self._format_data_networkx()
```
Comment (JGSweets) [inline dataprofiler/data_readers/graph_data.py:73]: Remove from here as this should occur if someone references the data auto. We only want to load the data if we need to load it :)

File: dataprofiler/data_readers/graph_data.py [hunk 2]
```diff
@@ -5,15 +5,73 @@
...
+
+        :param input_file_path: path to the file being loaded or None
+        :type input_file_path: str
+        :param data: data being loaded into the class instead of an input file
+        :type data: multiple types
+        :param options: options pertaining to the data type
+        :type options: dict
+        :return: None
+        """
+
```
[the remark this answers is not in the document]
Comment (taylorfturner) [reply dataprofiler/data_readers/graph_data.py:53]: @JGSweets need a `_check_and_return_options` specific to Graph Data? `_check_and_return_options` in CSVData would get run when instantiating the class

File: dataprofiler/data_readers/graph_data.py [hunk 3]
```diff
@@ -37,17 +95,14 @@ def _find_target_string_in_column(self, column_names, keyword_list):
                         break
             if has_target:
                 break
-        
         return target_index
 
-
     @classmethod
     def csv_column_names(cls, file_path, options):
```
Comment (JGSweets) [inline dataprofiler/data_readers/graph_data.py:101]: right now this presumes the first row is the header. Some CSVs have comments at the top. We cause use the info from `CSVData.is_match` and options to determine this. (will also need a test, e.g. single graph data file which has comments before the column names start).
to make this more generic we could do:
```python
@classmethod
def csv_column_names(cls, file_path, delimiter, quotechar, header): 
    column_names = []
    if header is None:
        return column_names

    for _ in range(header + 1):
        row = next(csv_reader)
    .
    .
    .
```
csv.reader can be updated to handle quotechar too:
```python
csv_reader = csv.reader(csv_file, delimiter=delimiter, quotechar=quotechar)
```

File: dataprofiler/data_readers/graph_data.py [hunk 4]
```diff
@@ -90,7 +143,37 @@ def is_match(cls, file_path, options=None):
             options.update(destination_node = destination_index)
             options.update(destination_list = target_keywords)
             options.update(source_list = source_keywords)
-            options.update(column_name = column_names)
+            options.update(column_names = column_names)
```
Comment (JGSweets) [inline dataprofiler/data_readers/graph_data.py:146]: nice!

File: dataprofiler/data_readers/graph_data.py [hunk 5]
```diff
@@ -90,7 +143,37 @@ def is_match(cls, file_path, options=None):
...
+        '''
+        Formats the input file into a networkX graph
+        '''
+        networkx_graph = nx.DiGraph()
+
+        # read lines from csv
+        csv_as_list = []
+        data_as_pd = data_utils.read_csv_df(self.input_file_path,self._delimiter,self._header,[],read_in_string=True,encoding=self.file_encoding)
+        data_as_pd = data_as_pd.apply(lambda x: x.str.strip())
+        csv_as_list = data_as_pd.values.tolist()
```
[the remark this answers is not in the document]
Comment (JGSweets) [reply dataprofiler/data_readers/graph_data.py:160]: Can we loop through the df?
or can we use:
https://networkx.org/documentation/stable/reference/generated/networkx.convert_matrix.from_pandas_edgelist.html

File: dataprofiler/tests/data_readers/test_csv_graph_data.py [hunk 6]
```diff
@@ -87,7 +85,39 @@ def test_is_graph_negative_1(self):
         Determine if the input CSV file can be automatically recognized as not being a graph w/ no options selected
         """
         for input_file in self.input_file_names_neg:
-            self.assertFalse(GraphData.is_match(input_file["path"]))   
+            self.assertFalse(GraphData.is_match(input_file["path"]))  
+
+    # test loading data
+    def test_data_loader_nodes(self):
```
[the remark this answers is not in the document]
Comment (JGSweets) [reply dataprofiler/tests/data_readers/test_csv_graph_data.py:91]: Maybe we can add the expected result of nodes, edges in the dict within the `setUpClass`?

File: dataprofiler/data_readers/graph_data.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (JGSweets) [reply dataprofiler/data_readers/graph_data.py:53]: @taylorfturner It would, but we aren't inheriting `CSVData`

**Review by taylorfturner — APPROVED** [review]

**Review by taylorfturner — APPROVED** [review]

**Review by JGSweets — APPROVED** [review]

**Pull request merged**